### DISCUSSION TITLE
Chunkserver and multiple interfaces

### DISCUSSION THREAD

**Discussion by alpharde** [moosefs/moosefs#505] [Q&A]
Is it possible to have some chunk servers to talk with each other through another interface? Some of them are directly connected on a dedicated interface.

**Comment by chogata** [comment]
It's not possible for only chunk servers. There is an option to route communication with master (metadata) through another interface than i/o communication, but i/o communication includes communication with clients (mounts).
That's of course what MooseFS allows you to do, there might be a way to do it using firewall rules.

If you wanted to use the option that is built-in in MooseFS, then you need to set corresponding IP numbers for the two interfaces in your chunk server machines. For example, if you have 9 chunk servers and 6 of them are connected via another set of interfaces, you can do something like:

CS1: first interface: 10.0.1.1, second interface: 10.0.2.1
CS2: first interface: 10.0.1.2, second interface: 10.0.2.2
CS3: first interface: 10.0.1.3, second interface: 10.0.2.3
CS4: first interface: 10.0.1.4, second interface: 10.0.2.4
CS5: first interface: 10.0.1.5, second interface: 10.0.2.5
CS6: first interface: 10.0.1.6, second interface: 10.0.2.6
CS7: first interface: 10.0.1.11
CS8: first interface: 10.0.1.12
CS9: first interface: 10.0.1.13

REMAP_BITS = 29
REMAP_SOURCE_IP_CLASS = 10.0.1.0
REMAP_DESTINATION_IP_CLASS =  10.0.2.0

**In this scenario ALL clients (mounts) MUST have access to 10.0.2.0 network, because the remapping will work for both chunk servers and clients.**

What this does: each time a module (client/chunkserver) asks the master for the address of a chunkserver it needs to talk with (to replicate a chunk, to write/read data to/from a chunk), the master, before sending the address, looks if maybe it needs to be re-mapped. So chunkservers connected to the master via 10.0.1.0/29 IP numbers will be presented to the asking module as a corresponding number from 10.0.2.0/29 class, the first 29 bits of the address will be re-mapped, the last 3 bits (0-7) will be unchanged. So all modules talk to the master on the 10.0.1.0 network, but the six first chunk servers will talk with each other on the 10.0.2.0 network AND all the existing clients will talk with the first six chunk servers on the 10.0.2.0 network.
